test/run-drun/par.mo

```diff
@@ -0,0 +1,152 @@
+
+        func closA(i : Nat) : async Nat {
+            assert Cycles.available() == 101;
+            i + message.size()
+        };
+
+        func closB() : async Nat = async {
+            assert Cycles.available() == 102;
+            message.size()
+        };
+
+        assert 42 == (await (with cycles = 101) closA(39));
+        assert 3 == (await (with cycles = 102) closB());
+
+        let c : async () =
+          (with yeah = 8; timeout = 55; cycles = 1000)
+          foo(func() : async () = async { assert message == "Hi!" and Cycles.available() == 3000 });
+        await c;
+        await (with cycles = 5000)
+        bar(func() : async () = async { assert message == "Hi!" and Cycles.available() == 4000 });
+    };
+
+    public func test2() : async () {
+        debugPrint "test2()";
+        await (with cycles = 1042) async { assert Cycles.available() == 1042 };
+    };
+
+    public func test3() : async () {
+        debugPrint "test3()";
+        oneshot();
+        (with cycles = 3456) oneshot();
+        (with timeout = 5) oneshot();
+        (with timeout = 5; cycles = 4567) A.oneshot();
+    };
+
+    public func test4() : async () {
+        debugPrint "test4()";
+        ignore await call_raw(principalOfActor A, "rawable", "DIDL\00\00");
+        ignore await (with cycles = 3456) call_raw(principalOfActor A, "rawable", "DIDL\00\00");
+    };
+
+    public func test5() : async () {
+        let t : async () =
+          (with timeout = 3; moot = "M") async {
+            debugPrint "test5()";
+            assert 0 != replyDeadline();
+          };
+        await t
+    };
+
+    public func ext() : async () {
+        assert 0 != replyDeadline();
+        debugPrint ("ext: " # debug_show(Cycles.available()));
+    };
+
+    public func test6() : async () {
+        await (with timeout = 3; cycles = 6543) A.ext()
+    };
+
+    public func test7() : async () {
+        debugPrint "test7()";
+        try
+          await (with timeout = 1) async {
+              assert 0 != replyDeadline();
+              // busy loop
+              loop ignore await (actor "aaaaa-aa" : actor { raw_rand : () -> async Blob }).raw_rand();
+              ()
+          }
+        catch e {
+            debugPrint("CAUGHT: " # debug_show errorCode e # " " # debug_show errorMessage e)
+        }
+    };
+
+    public func test8() : async () {
+        debugPrint "test8()";
+        await (with cycles = 98765) quux();
+        //await (with cycles = 87654) baz(); NOT YET!
+        //await (with cycles = 76543) A.baz() NOT YET!
+    };
+
+    func localOuter() : async () { debugPrint ("localOuter: " # debug_show(Cycles.available())) };
+    public func test9() : async () {
+        debugPrint "test9()";
+        var env = 42;
+        func local() : async () { debugPrint ("local: " # debug_show(env + Cycles.available())) };
+
+        // give a (dynamically) bogus base
+        object base { /*public*/ func cycles() : Text = "bogus" };
```

Comment:
Would `cycles` even be present in `base` if not declared public? I guess it depends on how we compile objects, but I suspect not.

Reply:
It would be present dynamically in `base`, but it won't enter into the parenthetical record, as it does only copy fields that are around in the type.

Reply:
With the new approach (imperative assignments to backend vars) we always follow the type, and never scan fields dynamically. So it doesn't really matter.